### PR TITLE
Include sender's public key in the encrypted message 

### PR DESCRIPTION
**Pull request by RoyTimes**
This fixes https://github.com/polkadot-js/common/issues/1314 and added the code posted by @LaurentTrk in the tests.

### CODE REVIEW

File: packages/keyring/src/pair/index.spec.ts [hunk 1]
```diff
@@ -171,20 +154,36 @@ describe('pair', (): void => {
...
     ).toEqual(message);
   });
 
+  it('passes the test on issue#1314', (): void => {
+    // Ref: https://github.com/polkadot-js/common/issues/1314
+
+    const keyring = new Keyring();
+
+    const senderPair = keyring.createFromUri(mnemonicGenerate(), { name: 'first pair' }, 'ed25519');
+    const receiverPair = keyring.createFromUri(mnemonicGenerate(), { name: 'second pair' }, 'ed25519');
```
Comment (LaurentTrk) [inline packages/keyring/src/pair/index.spec.ts:177]: Unfortunately, the issue #1314 talks about `sr25519`, the previous code was ok with `ed25519`.

Comment (RoyTimes) [reply packages/keyring/src/pair/index.spec.ts:177]: Uh. Sorry! Forget to change this in testing. Updating soon.

File: packages/keyring/src/pair/index.ts [hunk 2]
```diff
@@ -175,7 +175,7 @@ export function createPair ({ toSS58, type }: Setup, { publicKey, secretKey }: P
 
       const { nonce, sealed } = naclSeal(u8aToU8a(message), convertSecretKeyToCurve25519(secretKey), convertPublicKeyToCurve25519(u8aToU8a(recipientPublicKey)), nonceIn);
```
Comment (LaurentTrk) [inline packages/keyring/src/pair/index.ts:176]: This code, especially the `convertPublicKeyToCurve25519()` call, is likely to fail (depending on the bytes generated)

```javascript
    const sr25519Pair = keyring.createFromUri(mnemonicGenerate(), { name: 'an sr25519 pair' }, 'sr25519');
    const sr25519Curve = convertPublicKeyToCurve25519(u8aToU8a(sr25519Pair.publicKey));
    console.log(sr25519Curve);  //    <------ returns a null value
```

Comment (RoyTimes) [reply packages/keyring/src/pair/index.ts:176]: My mistake. I forgot to replace ed25519 to sr25519 in testing. I have done the testing and this fix is not correct yet.

**Comment by RoyTimes** [comment]
I introduced a new member of Keypair called `encryptionPublicKey` that is derived from `secretKey.slice(0, 32)` and the public key is calculated on Curve25519. I'm mostly certain that this is a secure operation but leaves the extension or application built on top of this harder, as you cannot simply use the address to send encrypted message anymore. Converting sr25519 pub key to curve25519 is risky (to implement). Doing ECDH directly on sr25519 is possible but harder than converting and is a nightmare to deal with and require users to be on the same KeyringType. So far, I can't think of a quicker and low-risk solution. 

Also, it seems to me that decode/encode wallet from secrets might be problematic that might worth attention, because the initial tests passed on cross-usage of ed25519/sr25519. On the tests proposed by @LaurentTrk the key conversion from `pairA.publicKey` to curve25519 should failed, because `ed2curve` cannot correctly calculate keys on sr25519. But it didn't fail on the original test. Therefore, I just assumed that it works. But apparently, I was wrong. Let me know if I'm, thinking about this the wrong way. 

```javascript
  it('validates encryption between ed25519 & sr25519 key pairs', (): void => {
    const PASS = 'testing';
    const encodedA = keyring.alice.encodePkcs8(PASS);
    const encodedB = keyring.bob.encodePkcs8(PASS);

    const pairA = createPair({ toSS58, type: 'sr25519' }, { publicKey: keyring.alice.publicKey });
    const pairB = createPair({ toSS58, type: 'sr25519' }, { publicKey: keyring.bob.publicKey });

    pairA.decodePkcs8(PASS, encodedA);
    pairB.decodePkcs8(PASS, encodedB);
    const message = new Uint8Array([0x61, 0x62, 0x63, 0x64, 0x65]);

    // alice-sr25519 -> bob-ed25519
    const encrypted1 = pairA.encryptMessage(message, keyring.bob.publicKey);

    // alice-sr25519 -> bob-sr25519
    const encrypted2 = pairA.encryptMessage(message, pairB.publicKey);

    // alice-ed25519 -> bob-sr25519
    const encrypted3 = keyring.alice.encryptMessage(message, pairB.publicKey);

    // decrypt: Bob-ed25519 - use alice-ed25519 pubkey
    expect(
      keyring.bob.decryptMessage(encrypted3, keyring.alice.publicKey)
    ).toEqual(message);

    // decrypt: Bob-ed25519 - use alice-sr25519 pubkey
    expect(
      keyring.bob.decryptMessage(encrypted1, pairA.publicKey)
    ).toEqual(message);

    // decrypt: Bob-sr25519 - use alice-sr25519 pubkey
    expect(
      pairB.decryptMessage(encrypted2, pairA.publicKey)
    ).toEqual(message);
  });
```

Also, I think it's now possible to do encryption/decryption on secp256k1 now if doing it this way. Will test it out this week.

**Comment by LaurentTrk** [comment]
Thanks @RoyTimes !  
As a non cryptographer, it sounds to me that we are trying to force a square peg into a round hole...  

I would prefer to go for a well documented approach like [ECIES](https://cryptobook.nakov.com/asymmetric-key-ciphers/ecies-public-key-encryption), which seems compatible both with [ed25519](https://github.com/phayes/ecies-ed25519) and sr25519 (using the `schnorrkelAgreement` already available in `wasm-crypto`).

In this approach, we only need one keypair (the one of the receiver), so we don't care about cross wallet exchange.

I can start to work on this, any thought/idea is welcome :)

**Comment by jacogr** [comment]
Sorry about the delay in looking, a bit behind on reviews - I have a note of getting to this one.

**Review by jacogr — COMMENTED** [review]
I do believe I prefer the newer approach in https://github.com/polkadot-js/common/pull/1331 since it does rely on a more standard approach to exchange information. 

It does also suffer from some "please communicate this information before using", in this case it would be the non-generally-available encrypted public key, in the other case it is the explicit keypair type, which is most probably a lesser evil.

I would prefer to converge on one solution between the two, intead of playing lord-and-master and having to decide myself between 2 different solutions for the same underlying issue.

**Comment by jacogr** [comment]
Closing, removing this functionality in https://github.com/polkadot-js/common/pull/1762

**Pull request closed without merging**